### ISSUE TITLE
M2M 100 for low resource languages

### ISSUE THREAD

**Issue by bmtm** [pytorch/fairseq#3234]
I'm trying to evaluate M2M 100 for some production tasks, but having issues for certain low resource languages like fr-wo. I'm getting good performance for high-resource pairs like en-zh so I'm just wondering if this is a known issue or if I'm doing something wrong.

For example:
```
fairseq-generate bin --batch-size 32 --path 1.2B_last_checkpoint.pt --fixed-dictionary model_dict.128k.txt -s fr -t wo --remove-bpe 'sentencepiece' --beam 5 --task translation_multi_simple_epoch --lang-pairs language_pairs_small_models.txt --decoder-langtok --encoder-langtok src --gen-subset test

S-2	__fr__ Maguire a émigré avec sa famille de l'Irlande au Maryland à l'âge de six ans, où il a étudié sous les jésuites au Collège Saint John, et est entré dans la Compagnie de Jésus en 1837.
H-2	-0.842350959777832	Maguire so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee so amee.

S-1	__fr__ SpaceX construit une fusée dans la région reculée de Boca Chica, située dans le comté de Cameron, au Texas, aux États-Unis.
H-1	-2.058246374130249	SpaceX biir rakett bi Boca Chica, biir Boca Chica, Cameron County, Texas, USA

S-3	__fr__ J'aime manger du melon
H-3	-4.000425815582275	Tél. : Tél.

S-0	__fr__ Je navigue sur Facebook tous les jours
H-0	-2.928954601287842	Lu Ma Mi Ju Vi Sa Do

S-4	__fr__ Je suis mongole
H-4	-4.857517719268799	Tembam
```
I've gotten similar results for these languages: wo uz tn ss sd or ns my ln lg kn kk gu ga ff br ast
ga was particularly interesting because the output tended to be NSFW.

anyways, I wanted to ask if
1. is this the expected result? (I chose fr-wo because it was specifically mentioned in the paper as having had human evalutation, but maybe those sentences were biased toward longer/more semantically dense input?)
2. is there any documentation on which translation directions are low resource? There's appendix A figure 13, but it says 100 languages and only displays 34. And this appears to only be the dictionary coverage and not training data.

**Comment by huihuifan** [comment]
Unfortunately, we acknowledge that many languages require significant work (see Discussion section). For languages with e.g. WMT/WAT focus, particularly we saw decreases in improvement in Czech, Burmese, and Chinese. There are a large variety of reasons, from language identification systems that require improvement (we used fasttext, which was trained on Wikipedia + Commoncrawl) producing monolingual data that is not in the correct language, to mining challenges (LASER was trained on 93 languages, but some languages have very little data), to low resource languages having poor generalization, and general necessity to improve our understanding of how to model under-resourced languages. Even backtranslation, for example, can be challenging if the original model is not high quality. 

Around your specific points - I agree that our fr-wo was quite generous, but given that I do not consider myself fluent in either language, I trusted their judgment. For ga, I will look into the challenges there - we did identify many sources that were not appropriate, but attempted to filter them out for most languages. I would guess there is a challenge in language identification here as well. For low resource languages, we are updating the paper with greater details for an update on arXiv, but essentially we tried to follow WMT20. In WMT20, Tamil and Inuktitut were both low-resource languages. Inuktitut had about 1million training sentences (including monolingual data for BT) and Tamil had 40M (including monolingual data for BT) - we decided to treat monolingual data similar to mined data given that BT and mining are both noisy sources of aligned data. So we considered anything with less than 50M mined training examples to be low resource. However, the languages you are mentioning in your list have far far below this, some even in the low thousands. 

Overall, I view this as an important area of progress - in our tables, I wanted to highlight that many important language directions have 0 BLEU (e.g. translating between different African languages) - because it's also relevant to indicate clearly what needs work. I hope that we can improve this, with much careful work, soon.
